vault/identity_store_identities.go

```diff
@@ -0,0 +1,342 @@
+	var err error
+	var newIdentity bool
+	var entity *entityStorageEntry
+	var previousEntity *entityStorageEntry
+
+	// Identity will be nil when a new identity is being registered; create a
+	// new struct in that case.
+	if identity == nil {
+		identity = &identityIndexEntry{}
+		newIdentity = true
+	}
+
+	// Get entity id
+	entityID := d.Get("entity_id").(string)
+	if entityID != "" {
+		entity, err := i.memDBEntityByID(entityID)
+		if err != nil {
+			return nil, err
+		}
+		if entity == nil {
+			return logical.ErrorResponse("invalid entity id"), nil
+		}
+	}
+
+	// Get identity name
+	identityName := d.Get("name").(string)
+	if identityName == "" {
+		return logical.ErrorResponse("missing identity name"), nil
+	}
+
+	// Get mount path to which the identity belongs to
+	mountPath := d.Get("mount_path").(string)
+	if mountPath == "" {
+		return logical.ErrorResponse("missing mount path"), nil
+	}
+
+	// Get identity metadata
+
+	// Accept metadata in the form of map[string]string to be able to index on
+	// it
+	var identityMetadata map[string]string
+	identityMetadataRaw, ok := d.GetOk("external_metadata")
+	if ok {
+		identityMetadataInput := identityMetadataRaw.(map[string]interface{})
+		if len(identityMetadataInput) > 0 {
+			identityMetadata = make(map[string]string, len(identityMetadataInput))
+			for k, v := range identityMetadataInput {
+				vStr, ok := v.(string)
+				if !ok {
+					return logical.ErrorResponse(fmt.Sprintf("invalid metadata value %#v for key %q", v, k)), nil
+				}
+				identityMetadata[k] = vStr
+			}
+		}
+	}
+
+	mountValidationResp := i.validateMountPathFunc(mountPath)
+	if mountValidationResp == nil {
+		return logical.ErrorResponse(fmt.Sprintf("invalid mount path %q", mountPath)), nil
+	}
+
+	identityByFactors, err := i.memDBIdentityByFactors(mountValidationResp.MountID, identityName)
+	if err != nil {
+		return nil, err
+	}
+
+	if newIdentity {
+		if identityByFactors != nil {
+			return logical.ErrorResponse("combination of mount path and identity name is already in use"), nil
+		}
+
+		// If this is an identity being tied to a non-existent entity, create
+		// a new entity for it.
+		if entity == nil {
+			entity = &entityStorageEntry{
+				Identities: []*identityIndexEntry{
+					identity,
+				},
+			}
+		} else {
+			entity.Identities = append(entity.Identities, identity)
+		}
+	} else {
+		// Verify that the combination of identity name and mount path is not
+		// already tied to a different identity
+		if identityByFactors != nil && identityByFactors.ID != identity.ID {
+			return logical.ErrorResponse("combination of mount path and identity name is already in use"), nil
+		}
+
+		// Fetch the entity to which the identity is tied to
+		existingEntity, err := i.memDBEntityByIdentityID(identity.ID)
+		if err != nil {
+			return nil, err
+		}
+
+		if existingEntity == nil {
+			return nil, fmt.Errorf("identity is not associated with an entity")
+		}
+
+		if entity != nil && entity.ID != existingEntity.ID {
```

Comment:
It feels early to do this block, because things later might return an error (during the sanitization functions). Should the identity be constructed first, then have this logic run directly before the upsert?